### PR TITLE
perf(parser): use memchr for lexing comments

### PR DESCRIPTION
**Pull request by sno2**
## Summary

This might improve the performance of lexing. Not sure, though.

## Test Plan

This was tested using the existing tests.

### CODE REVIEW

**Comment by sno2** [comment]
Improvements seem neglible: https://codspeed.io/astral-sh/ruff/branches/sno2:main

File: crates/ruff_python_parser/src/lexer.rs [hunk 1]
```diff
@@ -407,6 +407,22 @@ impl<'source> Lexer<'source> {
         #[cfg(debug_assertions)]
         debug_assert_eq!(self.cursor.previous(), '#');
 
+        #[cfg(any(
+            target_arch = "x86_64",
+            target_arch = "aarch64",
+            target_arch = "wasm32"
+        ))]
```
Comment (MichaReiser) [inline crates/ruff_python_parser/src/lexer.rs:414]: I think `memchr` handl this automatically and not having the configuration ensures that newly supported platforms automatically profit from a faster `memchr` routine.

**Review by MichaReiser — COMMENTED** [review]
It gives us a small speed up for the lexer. I didn't expect parsing to improve, because it performs way too many allocations. However, this could speed up things once we improve our parser.

File: crates/ruff_python_parser/src/lexer.rs (hunk 1 above)
Comment (sno2) [reply crates/ruff_python_parser/src/lexer.rs:414]: True, will look into this more.

**Comment by sno2** [comment]
I've given up on trying to optimize the String creation so this is ready for review :^) It seems that the only way we could get a performance boost from string allocation in the lexer is to not do it at all and use lifetimes and `&str`s.

**Comment by MichaReiser** [comment]
> I've given up on trying to optimize the String creation so this is ready for review :^) It seems that the only way we could get a performance boost from comment string allocation in the lexer is to not do it at all and use lifetimes and `&str`s.

I once had a pr that used `SmolStr` instead of `String` in the lexer to speed up things, but it didn't improve performance. I suspected that it was mainly because our string parsing is so slow (and allocates all over the place). It might be worth to give this another try after this improvement lands

**Review by MichaReiser — APPROVED** [review]
Thank you :)

**Pull request merged**